src/matmul.jl

```diff
@@ -30,39 +30,47 @@ function matmul!(C::AbstractMatrix{T}, A::AbstractMatrix{T}, B::AbstractMatrix{T
 
     GC.@preserve Amem begin
         α = T(_α);
-        for n ∈ StaticInt{1}():Nc:N # loop 5
-            nsize = min(Int(n + Nc), Int(N + 1)) - n
-            β = T(_β)
-            for k ∈ StaticInt{1}():Kc:K # loop 4
-                ksize = min(Int(k + Kc), Int(K + 1)) - k
-                Bview = view(B, k:k+ksize-1, n:n+nsize-1)
-                # seperate out loop 3, because of _Bblock type instability
-                if do_not_pack_B
-                    # _Bblock is likely to have the same type as _Bblock; it'd be nice to reduce the amount of compilation
-                    # by homogenizing types across branches, but for now I'm prefering the simplicity of using `Bview`
-                    # _Bblock = PointerMatrix(gesp1(stridedpointer(B), (k,n)), ksize, nsize)
-                    # matmul_loop3!(C, Aptr, Ablock, A, _Bblock, α, β, msize, ksize, nsize, M, k, n, Mc)
-                    matmul_loop3!(C, Aptr, A, Bview, α, β, ksize, nsize, M, k, n, Mc)
-                else
-                    Bblock = PointerMatrix(Bptr, (ksize,nsize))
-                    unsafe_copyto_avx!(Bblock, Bview)
-                    matmul_loop3!(C, Aptr, A, Bblock, α, β, ksize, nsize, M, k, n, Mc)
+        full_range = StaticInt{1}():Nc:N
+        partitioned_range = Iterators.partition(full_range, BLIS_JC_NT[])
+        @_sync for partition ∈ partitioned_range
+            @_spawn for n ∈ partition # loop 5
+                nsize = min(Int(n + Nc), Int(N + 1)) - n
+                β = T(_β)
+                for k ∈ StaticInt{1}():Kc:K # loop 4
+                    ksize = min(Int(k + Kc), Int(K + 1)) - k
+                    Bview = view(B, k:k+ksize-1, n:n+nsize-1)
+                    # seperate out loop 3, because of _Bblock type instability
+                    if do_not_pack_B
+                        # _Bblock is likely to have the same type as _Bblock; it'd be nice to reduce the amount of compilation
+                        # by homogenizing types across branches, but for now I'm prefering the simplicity of using `Bview`
+                        # _Bblock = PointerMatrix(gesp1(stridedpointer(B), (k,n)), ksize, nsize)
+                        # matmul_loop3!(C, Aptr, Ablock, A, _Bblock, α, β, msize, ksize, nsize, M, k, n, Mc)
+                        matmul_loop3!(C, Aptr, A, Bview, α, β, ksize, nsize, M, k, n, Mc)
+                    else
+                        Bblock = PointerMatrix(Bptr, (ksize,nsize))
+                        unsafe_copyto_avx!(Bblock, Bview)
+                        matmul_loop3!(C, Aptr, A, Bblock, α, β, ksize, nsize, M, k, n, Mc)
```

Comment:
Unless I'm missing something, `Bptr` here is the same across threads. Meaning each thread will be trying to read from and write to the same block of memory.

Reply:
That would explain the failure when you multithread loop 5, but I'm not sure about why you'd have a problem when threading loop3.